cpp/src/sampling/nbr_sampling.cuh

```diff
@@ -0,0 +1,543 @@
+        d_acc_indices.resize(new_sz, handle.get_stream());
+
+        // zip quad; must be done after resizing,
+        // because they grow from one iteration to another,
+        // so iterators could be invalidated:
+        //
+        auto acc_zip_it =
+          thrust::make_zip_iterator(thrust::make_tuple(d_acc_src.begin() + old_sz,
+                                                       d_acc_dst.begin() + old_sz,
+                                                       d_acc_ranks.begin() + old_sz,
+                                                       d_acc_indices.begin() + old_sz));
+
+        // union step:
+        //
+        auto out_zip_it = thrust::make_zip_iterator(thrust::make_tuple(
+          d_out_src.begin(), d_out_dst.begin(), d_out_ranks.begin(), d_indices.begin()));
+
+        thrust::copy_n(handle.get_thrust_policy(), out_zip_it, add_sz, acc_zip_it);
+
+        // shuffle step: update input for self_rank
+        // zipping is necessary to preserve rank info during shuffle!
+        //
+        auto next_in_zip_begin =
+          thrust::make_zip_iterator(thrust::make_tuple(d_out_dst.begin(), d_out_ranks.begin()));
+        auto next_in_zip_end =
+          thrust::make_zip_iterator(thrust::make_tuple(d_out_dst.end(), d_out_ranks.end()));
+
+        update_input_by_rank(handle,
+                             graph_view,
+                             next_in_zip_begin,
+                             next_in_zip_end,
+                             static_cast<size_t>(self_rank),
+                             d_in,
+                             d_ranks,
+                             gpu_t{});
+      }
+
+      //}
+      ++level;
+    }
+
+    return std::make_tuple(
+      std::move(d_acc_src), std::move(d_acc_dst), std::move(d_acc_ranks), std::move(d_acc_indices));
+  } else {
+    CUGRAPH_FAIL("Neighborhood sampling functionality is supported only for the multi-gpu case.");
+  }
+}
+
+}  // namespace detail
+
+/**
+ * @brief Multi-GPU Uniform Neighborhood Sampling.
+ * @tparam graph_view_t Type of graph view.
+ * @tparam index_t Type used for indexing; typically edge_t
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Graph View object to generate NBR Sampling on.
+ * @param ptr_d_start Device array of pairs: (starting_vertex_index, rank) for the NBR Sampling.
+ * @param num_starting_vs size of starting vertex set
+ * @param h_fan_out vector of branching out (fan-out) degree per source vertex for each level
+ * parameter used for obtaining local out-degree information
+ * @param flag_replacement boolean flag specifying if random sampling is done without replacement
+ * (true); or, with replacement (false); default = true;
+ * @return tuple of tuple of device vectors and counts:
+ * ((vertex_t source_vertex, vertex_t destination_vertex, int rank, edge_t index), rx_counts)
+ */
```

Comment:
Should we really duplicate documentations (in the public header and here as well.... we should put extra-effort to keep this up-to-date and in-sync with the documentation in the public header).

Reply:
So, what's the desired course of action here: _no_ documentation? some documentation...? I'll remove it for now.

Reply:
Generally I think the goal is to document functions in the place where they are defined.  If the function is implemented somewhere else then we don't want to replicate the documentation of the definition, since that leads to more potential for the documentation becoming inconsistent over time.  With our templating, we sometimes have a function defined in one place and implemented in multiple other places.  It just becomes a maintenance challenge if the same documentation is in multiple places.